Userland/Libraries/LibCompress/Deflate.cpp

```diff
@@ -448,4 +464,571 @@ void DeflateDecompressor::decode_codes(CanonicalCode& literal_code, Optional<Can
+            previous_match_length = match_length;
+
+            if (match_length == maximum_match_length)
+                return match_length; // bail if we got the maximum possible length
+        }
+
+        candidate = m_hash_prev[candidate % window_size];
+    }
+    if (!match_found)
+        return 0;                 // we didnt find any matches
+    return previous_match_length; // we found matches, but they were at most previous_match_length long
+}
+
+ALWAYS_INLINE u8 DeflateCompressor::distance_to_base(u16 distance)
+{
+    return (distance <= 256) ? distance_to_base_lo[distance - 1] : distance_to_base_hi[(distance - 1) >> 7];
+}
+
+template<size_t Size>
+void DeflateCompressor::generate_huffman_lengths(Array<u8, Size>& lengths, const Array<u16, Size>& frequencies, size_t max_bit_length)
+{
+    VERIFY((1u << max_bit_length) >= Size);
+    u16 heap_keys[Size]; // Used for O(n) heap construction
+    u16 heap_values[Size];
+
+    u16 huffman_links[Size * 2 + 1] = { 0 };
+    u16 frequency_cap = UINT16_MAX;
+try_again:
+    size_t non_zero_freqs = 0;
+    for (size_t i = 0; i < Size; i++) {
+        auto frequency = frequencies[i];
+        if (frequency == 0)
+            continue;
+
+        if (frequency > frequency_cap) {
+            frequency = frequency_cap;
+        }
+
+        heap_keys[non_zero_freqs] = frequency;               // sort symbols by frequency
+        heap_values[non_zero_freqs] = Size + non_zero_freqs; // huffman_links "links"
+        non_zero_freqs++;
+    }
+
+    // special case for only 1 used symbol
+    if (non_zero_freqs < 2) {
+        for (size_t i = 0; i < Size; i++)
+            lengths[i] = (frequencies[i] == 0) ? 0 : 1;
+        return;
+    }
+
+    BinaryHeap<u16, u16, Size> heap { heap_keys, heap_values, non_zero_freqs };
+
+    // build the huffman tree - binary heap is used for efficient frequency comparisons
+    while (heap.size() > 1) {
+        u16 lowest_frequency = heap.peek_min_key();
+        u16 lowest_link = heap.pop_min();
+        u16 second_lowest_frequency = heap.peek_min_key();
+        u16 second_lowest_link = heap.pop_min();
+
+        u16 new_link = heap.size() + 2;
+
+        heap.insert(lowest_frequency + second_lowest_frequency, new_link);
+
+        huffman_links[lowest_link] = new_link;
+        huffman_links[second_lowest_link] = new_link;
+    }
+
+    non_zero_freqs = 0;
+    for (size_t i = 0; i < Size; i++) {
+        if (frequencies[i] == 0) {
+            lengths[i] = 0;
+            continue;
+        }
+
+        u16 link = huffman_links[Size + non_zero_freqs];
+        non_zero_freqs++;
+
+        size_t bit_length = 1;
+        while (link != 2) {
+            bit_length++;
+            link = huffman_links[link];
+        }
+
+        if (bit_length > max_bit_length) {
+            VERIFY(frequency_cap != 1);
+            frequency_cap /= 2;
+            goto try_again; // FIXME: gotos are ugly, but i cant think of a good way to flatten this
```

Comment:
Sorry about that :-p

I think one way would be if generate_huffman_lengths took frequency_cap as a parameter (with default value UINT16_MAX). Then you could tail call here:

return generate_huffman_lengths(lengths, frequencies, max_bit_length, frequency_cap / 2)